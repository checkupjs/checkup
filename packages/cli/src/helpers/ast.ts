import * as recast from 'recast';
import * as t from '@babel/types';

import traverse, { TraverseOptions } from '@babel/traverse';

/**
 * @class AstTransformer
 *
 * @example
 *
 * let code = new AstTransformer(source)
 *              .traverse(visitors)
 *              .generate();
 */
export default class AstTransformer {
  ast!: t.File;
  constructor(public source: string) {
    this.ast = this._parse(source);
  }
  private _parse(source: string): t.File {
    return recast.parse(source, { parser: require('recast/parsers/typescript') });
  }
  traverse(visitors: TraverseOptions): AstTransformer {
    traverse(this.ast, visitors);
    return this;
  }
  generate(): string {
    return recast.print(this.ast, { quote: 'single', wrapColumn: 100 }).code;
  }
}
