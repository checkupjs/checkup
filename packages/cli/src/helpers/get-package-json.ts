import * as fs from 'fs-extra';
import * as path from 'path';

import { PackageJson } from 'type-fest';

/**
 * @param basePath
 */
export function getPackageJson(basePath: string, pathName: string = 'package.json'): PackageJson {
  let package_ = {};
  let packageJsonPath = path.join(path.resolve(basePath), pathName);

  try {
    package_ = fs.readJsonSync(packageJsonPath);
  } catch (error) {
    if (error.code === 'ENOENT') {
      throw new Error(
        `The ${path.resolve(
          basePath
        )} directory found through the 'path' option does not contain a package.json file. You must run checkup in a directory with a package.json file.`
      );
    }
  }

  return package_;
}
