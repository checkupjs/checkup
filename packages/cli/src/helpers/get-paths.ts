import { join, resolve } from 'path';
import { existsSync } from 'fs';

const micromatch = require('micromatch');
const isValidGlob = require('is-valid-glob');
const walkSync = require('walk-sync');
const globby = require('globby');

const PATHS_TO_IGNORE = [
  '**/node_modules/**',
  'bower_components/**',
  '**/tests/dummy/**',
  'concat-stats-for/**',
  'dist/**',
  'build/**',
  'vendor/**',
  '.git/**',
];

/**
 * @param basePath
 */
export function getFilePaths(
  basePath: string,
  globs: string[] = [],
  excludePaths: string[] = []
): string[] {
  let mergedPathsToIgnore = [...excludePaths, ...PATHS_TO_IGNORE];

  if (globs.length > 0) {
    return expandFileGlobs(globs, basePath, mergedPathsToIgnore);
  }
  const allFiles: string[] = walkSync(basePath, {
    ignore: mergedPathsToIgnore,
    directories: false,
  });

  return resolveFilePaths(allFiles, basePath);
}

function expandFileGlobs(
  filePatterns: string[],
  basePath: string,
  excludePaths: string[]
): string[] {
  return filePatterns.flatMap((pattern) => {
    let isLiteralPath = !isValidGlob(pattern) && existsSync(pattern);

    if (isLiteralPath) {
      let isIgnored = !micromatch.isMatch(pattern, PATHS_TO_IGNORE);

      if (!isIgnored) {
        return pattern;
      }

      return [];
    }

    let expandedGlobs = globby.sync(pattern, {
      cwd: basePath,
      ignore: excludePaths,
      gitignore: true,
    }) as string[];

    return resolveFilePaths(expandedGlobs, basePath);
  });
}

function resolveFilePaths(filePaths: string[], basePath: string): string[] {
  if (basePath !== '.') {
    return filePaths.map((pathName: string) => {
      return join(resolve(basePath), pathName);
    });
  }
  return filePaths;
}
