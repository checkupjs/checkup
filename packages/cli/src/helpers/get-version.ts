const { version } = require('../../package.json');

export function getVersion(fakeVersion: string = '0.0.0') {
  if (process.env.JEST_WORKER_ID !== undefined) {
    return fakeVersion;
  }

  return version;
}
