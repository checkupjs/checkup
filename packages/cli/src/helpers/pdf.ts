import * as Handlebars from 'handlebars';
import * as fs from 'fs';
import * as path from 'path';

import { pathToFileURL } from 'url';
import { printToPDF } from './print-to-pdf';
import { readFileSync } from 'fs-extra';

import tmp = require('tmp');

const date = require('date-and-time');
const REPORT_PATH = path.join(__dirname, '../static/report-template.hbs');
const CARD_PATH = path.join(__dirname, '../static/components/card.hbs');

let REPORT_TEMPLATE_RAW = readFileSync(REPORT_PATH, 'utf8');
let CARD_TEMPLATE_RAW = readFileSync(CARD_PATH, 'utf8');
/**
 * @param jsonResult
 * @param results
 */
export async function generateReport(
  resultOutputPath: string,
  resultsForPdf: any
): Promise<string> {
  let reportHTML = generateHTML(resultsForPdf);

  let tmpDir = fs.realpathSync(tmp.dirSync({ unsafeCleanup: true }).name);
  let htmlTmpPath = path.resolve(path.join(tmpDir, 'tmp-report.html'));
  fs.writeFileSync(htmlTmpPath, reportHTML);

  if (!fs.existsSync(resultOutputPath)) {
    fs.mkdirSync(resultOutputPath, { recursive: true });
  }

  let outputFilePath = path.resolve(
    path.join(
      resultOutputPath,
      `checkup-report-${date.format(new Date(), 'YYYY-MM-DD-HH-mm-ss')}.pdf`
    )
  );

  await printToPDF(pathToFileURL(htmlTmpPath).toString(), outputFilePath);

  return outputFilePath;
}

export function generateHTML(resultsForPdf: any) {
  const template = Handlebars.compile(REPORT_TEMPLATE_RAW);
  Handlebars.registerPartial('cardPartial', CARD_TEMPLATE_RAW);
  return template(resultsForPdf);
}
