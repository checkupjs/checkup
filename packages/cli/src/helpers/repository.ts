import { RepositoryInfo } from '../types';

const hash = require('promise.hash.helper');
const util = require('util');
const execAsPromise = util.promisify(require('child_process').exec);

const COMMIT_COUNT = "git log --oneline $commit | wc -l | tr -d ' '";
const FILE_COUNT = "git ls-files | wc -l | tr -d ' '";
const REPO_AGE =
  'git log --reverse --pretty=oneline --format=" % ar" | head -n 1 | LC_ALL=C sed \'s/ago//\'';
const ACTIVE_DAYS = `git log --pretty='format: %ai' $1 | cut -d ' ' -f 2 | sort -r | uniq | awk '
    { sum += 1 }
    END { print sum }
  '`;

/**
 * @param cmd
 * @param options
 * @param defaultValue
 * @param toType
 */
async function exec(
  cmd: string,
  options: any,
  defaultValue: string | number,
  toType: Function = String
) {
  const { stdout } = await execAsPromise(cmd, options);

  return toType(stdout.trim()) || defaultValue;
}

/**
 * @param path
 */
export function getRepositoryInfo(path: string): Promise<RepositoryInfo> {
  return hash({
    totalCommits: exec(COMMIT_COUNT, { cwd: path }, 0, Number),
    totalFiles: exec(FILE_COUNT, { cwd: path }, 0, Number),
    age: exec(REPO_AGE, { cwd: path }, '0 days'),
    activeDays: exec(ACTIVE_DAYS, { cwd: path }, '0 days'),
  });
}
