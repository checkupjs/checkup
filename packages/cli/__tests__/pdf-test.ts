import {
  Category,
  NumericalCardData,
  Priority,
  TaskMetaData,
  PieChartData,
  UIReportData,
} from '@checkup/core';

import { generateHTML } from '../src/helpers/pdf';

describe('generateHTML', () => {
  const taskMeta: TaskMetaData = {
    taskName: 'mock-task',
    friendlyTaskName: 'Mock Task',
    taskClassification: {
      category: Category.Migrations,
      priority: Priority.Medium,
    },
  };

  const projectMeta = {
    project: {
      name: 'some-app',
      version: '0.0.1',
      repository: {
        totalCommits: 5870,
        totalFiles: 1571,
        age: '8 years',
        activeDays: '1380',
      },
    },
    checkup: {
      configHash: '07e8f7d8731ffbb321ad86f8f2f62460',
      version: '0.0.6',
    },
  };

  const mergedResults: UIReportData = {
    meta: projectMeta,
    results: {
      [Category.Insights]: {
        [Priority.High]: [],
        [Priority.Medium]: [new NumericalCardData(taskMeta, 100, 'bad patterns in your app')],
        [Priority.Low]: [],
      },
      [Category.Migrations]: {
        [Priority.High]: [],
        [Priority.Medium]: [],
        [Priority.Low]: [],
      },
      [Category.Recommendations]: {
        [Priority.High]: [],
        [Priority.Medium]: [],
        [Priority.Low]: [],
      },
    },
    requiresChart: false,
  };

  it('returns correct HTML string', async () => {
    const htmlString = await generateHTML(mergedResults);
    expect(htmlString).toMatchInlineSnapshot(`
      "<html class=\\"text-gray-900 antialiased leading-tight\\">

      <head>
        <meta charset=\\"utf-8\\">
        <title>Checkup report</title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
        <link href=\\"https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css\\" rel=\\"stylesheet\\">
        <link rel=\\"stylesheet\\" href=\\"https://cdn.jsdelivr.net/npm/@tailwindcss/ui@latest/dist/tailwind-ui.min.css\\">
      </head>
      <body class=\\"min-h-screen bg-gray-100 p-5\\">
        <div class=\\"py-4 flex flex-wrap flex-col\\">
          <div class=\\"self-end text-xs\\">
            Checkup version 0.0.6
            <br />
            Config hash 07e8f7d8731ffbb321ad86f8f2f62460
          </div>
          <h1 class=\\"text-4xl\\">Checkup Report for some-app <span class=\\"text-xl\\">(v. 0.0.1)</span></h1>
          <h2 class=\\"text-base italic\\">This project is 8 years old, has been active for 1380 days, has 1571 files, and 5870 commits</h2>
        </div>

          <section class=\\"max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 border border-gray-500 m-5\\">
          <h3 class=\\"text-center capitalize text-3xl font-bold\\">insights</h3>
              <h4 class=\\"capitalize text-2xl font-bold m-5\\">medium priority insightss</h4>
                <div class=\\"grid grid-cols-2 gap-4\\">
                    <div class=\\"border rounded overflow-hidden shadow-lg bg-white my-4 row-start-1\\">
                      <div class=\\"px-6 py-4 flex flex-wrap flex-col\\">
                        <div class=\\"text-xl font-bold self-end\\">A</div>
                        <h1 class=\\"text-xl self-center\\">Mock Task</h1>
                        <div class=\\"text-4xl self-center\\">100</div>
                        <p>bad patterns in your app</p>
                      </div>
                    </div>
                </div>
          </section>
          <section class=\\"max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 border border-gray-500 m-5\\">
          <h3 class=\\"text-center capitalize text-3xl font-bold\\">migrations</h3>
          </section>
          <section class=\\"max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 border border-gray-500 m-5\\">
          <h3 class=\\"text-center capitalize text-3xl font-bold\\">recommendations</h3>
          </section>

      </body>
      </html>
      "
    `);
  });

  it('includes tailwind UI lib and stylesheet, but no chartjs (since results dont require it by default)', async () => {
    const htmlString = await generateHTML(mergedResults);

    expect(htmlString).toContain('tailwind.min.css');
    expect(htmlString).toContain('tailwind-ui.min.css');
    expect(htmlString).not.toContain('Chart.js v2.9.3 CSS');
    expect(htmlString).not.toContain('Chart.js v2.9.3 JS');
  });

  it('includes chartsjs css and js when there is a pie-chart being rendered as part of the results', async () => {
    const htmlString = await generateHTML({
      meta: projectMeta,
      results: {
        [Category.Insights]: {
          [Priority.High]: [],
          [Priority.Medium]: [],
          [Priority.Low]: [],
        },
        [Category.Migrations]: {
          [Priority.High]: [],
          [Priority.Medium]: [],
          [Priority.Low]: [],
        },
        [Category.Recommendations]: {
          [Priority.High]: [],
          [Priority.Medium]: [
            new PieChartData(
              {
                taskName: 'mock-task',
                friendlyTaskName: 'Mock Task',
                taskClassification: {
                  category: Category.Insights,
                  priority: Priority.Medium,
                },
              },
              [
                { value: 33, description: 'blah' },
                { value: 23, description: 'blah' },
                { value: 13, description: 'black' },
                { value: 3, description: 'sheep' },
              ],
              'this is a chart'
            ),
          ],
          [Priority.Low]: [],
        },
      },
      requiresChart: true,
    });

    expect(htmlString).toContain('Chart.js v2.9.3 CSS');
    expect(htmlString).toContain('Chart.js v2.9.3 JS');
  });
});
