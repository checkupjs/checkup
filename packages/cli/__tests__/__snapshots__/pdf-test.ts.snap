// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateHTML returns correct HTML string 1`] = `
"<html class=\\"text-gray-900 antialiased leading-tight\\">

<head>
  <meta charset=\\"utf-8\\">
  <title>Checkup report</title>
  <meta name=\\"description\\" content=\\"\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  <link href=\\"https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css\\" rel=\\"stylesheet\\">
  <link rel=\\"stylesheet\\" href=\\"https://cdn.jsdelivr.net/npm/@tailwindcss/ui@latest/dist/tailwind-ui.min.css\\">
</head>
<body class=\\"min-h-screen bg-gray-100\\">
  <h1>Checkup report</h1>

</body>
</html>
"
`;
